Revamp `sending emails from a service` tutorial (#8133)
ballerina-platform/ballerina-dev-website: swan-lake/integration-tutorials/sending-emails-from-a-service.md

Reviewer: > Extract necessary details from the request (e.g., hospital, patient, doctor, etc.) and make a call to the hospital backend service 

Can we still say that we are extracting `details from the request (e.g., hospital, patient, doctor, etc.)`?

    ```json
    {

>         "patientWithCardNo": {

Reviewer 2: This shouldn't change. Can still be patient.

1. Create a new Ballerina project using the `bal` command and open it in VS Code.

    ```
    $ bal new sending-emails-from-a-service
    ```

2. Remove the generated content in the `main.bal` file and open the diagram view in VS Code.

    ![Open diagram view](/learn/images/integration-tutorials/sending-emails-from-a-service/open_diagram_view.gif)

> 3. Generate a record type corresponding to the payload from the backend service using the ["Paste JSON as record"](https://wso2.com/ballerina/vscode/docs/references/convert-json-to-records/#via-the-command-palette) VS Code command by providing a sample of the expected JSON payload.

Reviewer: ```suggestion
3. Generate a record type corresponding to the payload from the backend service by providing a sample of the expected JSON payload.
```

We no longer using `Paste JSON as a record`

>     The payload from the backend service will be an array of JSON objects, where each JSON object will be similar to the following.

Reviewer: We are not getting an array right?

>     The payload from the backend service will be an array of JSON objects, where each JSON object will be similar to the following.

Reviewer 2: Is this correct? Not an array?

     
    - Define an HTTP resource that allows the `POST` operation on the resource path `/categories/{category}/reserve`, where `category` (corresponding to the specialization) is a path parameter.

    - Use `ReservationRequest` as a parameter indicating that the resource expects a JSON object corresponding to `ReservationRequest` as the payload. 

    - Use `http:Created|http:NotFound|http:InternalServerError` as the return type to indicate that the response will be an `http:Created` response when the email is sent successfully to the user or the response will be an `http:NotFound` or `http:InternalServerError` response on error.

        ![Define the resource](/learn/images/integration-tutorials/sending-emails-from-a-service/define_a_resource.gif)

        The generated service will be as follows.

        ```ballerina
        service /healthcare on new http:Listener(8290) {
            resource function post categories/[string category]/reserve(ReservationRequest payload) 
                    returns http:Created|http:NotFound|http:InternalServerError {
                
            }
        }
        ```

> 5. Define a [configurable variable](https://ballerina.io/learn/by-example/#configurability) for the URL of the backend service.

Reviewer: Shall we say define  "configurable variables" here but only show one is int the demo?

We can remove the "similarly.." part.

    > **Note:** Enable two factor authentication on your Google account, generate an app password, and use the app password in place of your email password. The app password can be generated with [this link](https://myaccount.google.com/apppasswords?pli=1&rapt=AEjHL4Mf5XDD4rE79YJpP5E2NoNwhvXMET_TWyBcQRn-HMzt0PI8BmptpMGRiBVIamW-0ECgVZtXxMRA19bL4Wfnq_hmjBEMqA).

> 

Reviewer 2: Shall we move this to after the definition? Kinda breaks the flow IMO.

>     ![Define a configurable variable](/learn/images/integration-tutorials/sending-a-message-to-a-service/define_a_configurable_variable.gif)

Reviewer: The gif is not working

    Similarly define configurable variables for the payment backend and host, username and password of the SMTP client.

    The generated code will be as follows.

    ```ballerina
    configurable string hospitalServicesBackend = "http://localhost:9090";
    configurable string paymentBackend = "http://localhost:9090/healthcare/payments";
    configurable string host = "smtp.gmail.com";
    configurable string username = ?;
    configurable string password = ?;
    ```

6. Define an [`http:Client`](https://ballerina.io/learn/by-example/#http-client) object to send requests to the backend service.
   

>    ![Define the clients](/learn/images/integration-tutorials/sending-emails-from-a-service/define_clients.gif)

Reviewer: We may have to crop in the middle of the video. But let's wait for other reviews and then do.

>    Similarly define another [`http:Client`](https://ballerina.io/learn/by-example/#http-client) object to send requests to the payment backend service and an [`email:SmtpClient`](https://ballerina.io/learn/by-example/#email-client) object to send emails to the user.

Reviewer: Shall remove this "similarly.." part. Doing a demo for one client is okay.

        int appointmentNumber;
        Doctor doctor;
        Patient patient;
        decimal fee;
        boolean confirmed;
        string card_number;
    |};

    type Payment record {|
        int appointmentNo;
        string doctorName;
        string patient;
        decimal actualFee;
        int discount;
        decimal discounted;
        string paymentID;
        string status;
    |};

    type ReservationRequest record {|
        record {|
            *Patient;
            string cardNo;
        |} patient;
        string doctor;
        string hospital_id;
        string hospital;
        string appointment_date;
    |};

    type ReservationResponse record {|
        int appointmentNo;
        string doctorName;
        string patient;
        decimal actualFee;
        int discount;
        decimal discounted;
        string paymentID;
        string status;
    |};
    ```
    - Use `http:Created|http:NotFound|http:InternalServerError` as the return type to indicate that the response will be an `http:Created` response when the email is sent successfully to the user or the response will be an `http:NotFound` or `http:InternalServerError` response on error.

>    - The `ReservationRequest` record uses [record type inclusion](https://ballerina.io/learn/by-example/type-inclusion-for-records/) in the type of the `patient` field to include all the fields from the `Patient` record along with the `cardNo` field.

Reviewer 2: Combine these 3 similar to Please change the flow similar to https://github.com/ballerina-platform/ballerina-dev-website/blob/integration-tutorials/swan-lake/integration-tutorials/sending-a-message-to-a-service.md?plain=1#L69 since they correspond to the GIF after.

> 5. Define [configurable variables](https://ballerina.io/learn/by-example/#configurability) for the port on which the listener should listen and the URLs of the backend services. Also, define configurable variables for the host, username, and password of the SMTP client.

Reviewer 2: We don't have a configurable variable for port anymore, do we?

Author: No we don't, removed it

>     ![Define a configurable variable](/learn/images/integration-tutorials/sending-a-message-to-a-service/define_a_configurable_variable.gif)

Reviewer 2: Still broken.

Reviewer: The location of the GIF is wrong. (Please check the path)

> 6. Define two [`http:Client`](https://ballerina.io/learn/by-example/#http-client) clients to send requests to the backend services and one [`email:SmtpClient`](https://ballerina.io/learn/by-example/#email-client) client to send emails.

Reviewer 2: Is it a tooling issue that we have to type in the url instead of being able to pick from suggestions? Please create an issue if so.

cc @dulajdilshan

Reviewer: Yeah. Tooling support needs to be improved by allowing haveing suggestions. Better to track via a GH isue

    $ bal new sending-emails-from-a-service
    ```

2. Remove the generated content in the `main.bal` file and open the diagram view in VS Code.

    ![Open diagram view](/learn/images/integration-tutorials/sending-emails-from-a-service/open_diagram_view.gif)

> 3. Generate a record type corresponding to the payload from the backend service by providing a sample of the expected JSON payload.

Reviewer 2: Since there's two let's make this plural.

```suggestion
3. Generate record types corresponding to the payloads from the backend services by providing samples of the expected JSON payloads.
```

>     The payload from the backend service will be a JSON object similar to the following.

Reviewer 2: Mention which backend.

    ```json
    {
        "appointmentNumber": 1,
        "doctor": {
            "name": "thomas collins",
            "hospital": "grand oak community hospital",
            "category": "surgery",
            "availability": "9.00 a.m - 11.00 a.m",
            "fee": 7000.0
        },
        "patient": {
            "name": "John Doe",
            "dob": "1940-03-19",
            "ssn": "234-23-525",
            "address": "California",
            "phone": "8770586755",
            "email": "johndoe@gmail.com"
        },
        "hospital": "grand oak community hospital",
        "confirmed": false,
        "appointmentDate": "2023-10-02"
    }
    ```

    ![Define records](/learn/images/integration-tutorials/sending-emails-from-a-service/define_records.gif)

> 

Reviewer 2: Add the generated record types wherever we say generate a record type. Similar to other snippets we've added.

@dulajdilshan, please note for the first tutorial also.

Reviewer: Ack

>     Similarly, generate record types for other payloads from the backend service.

Reviewer 2: Mention which backend specifically, since there are only two.

Author: Removed the `Similarly..` part

    We can use [record type inclusion](https://ballerina.io/learn/by-example/type-inclusion-for-records/) for the `ReservationRequest` record type to include the `Patient` record along with the `cardNo` field within the type of the `patient` field.

    > **Note:**
    > While it is possible to work with the JSON payload directly, using record types offers several advantages including enhanced type safety, data validation, and better tooling experience (e.g., completion).

4. Define the [HTTP service (REST API)](https://ballerina.io/learn/by-example/#rest-service) that has the resource that accepts user requests, retrieves relevant details from the backend service, and responds to the request.

>     - Open the [Ballerina HTTP API Designer](https://wso2.com/ballerina/vscode/docs/design-the-services/http-api-designer) in VS Code.

Reviewer 2: We need to avoid anonymous records. Is the source up to date with the latest changes? Offline, you showed a separate record type `PatientWithCardNo`?

We need to extract the anonymous record and use inclusion there, not in `ReservationRequest`.

                
            }
        }
        ```

5. Define [configurable variables](https://ballerina.io/learn/by-example/#configurability) for the URLs of the backend services. Also, define configurable variables for the host, username, and password of the SMTP client.

    ![Define a configurable variable](/learn/images/integration-tutorials/sending-emails-from-a-service/define_a_configurable_variable.gif)

    The generated code will be as follows.

    ```ballerina
    configurable string hospitalServicesBackend = "http://localhost:9090";
    configurable string paymentBackend = "http://localhost:9090/healthcare/payments";
    configurable string host = "smtp.gmail.com";
    configurable string username = ?;
    configurable string password = ?;
    ```

    > **Note:** Enable two factor authentication on your Google account, generate an app password, and use the app password in place of your email password. The app password can be generated with [this link](https://myaccount.google.com/apppasswords?pli=1&rapt=AEjHL4Mf5XDD4rE79YJpP5E2NoNwhvXMET_TWyBcQRn-HMzt0PI8BmptpMGRiBVIamW-0ECgVZtXxMRA19bL4Wfnq_hmjBEMqA).

> 6. Define two [`http:Client`](https://ballerina.io/learn/by-example/#http-client) clients to send requests to the backend services and one [`email:SmtpClient`](https://ballerina.io/learn/by-example/#email-client) client to send emails.

Reviewer 2: ```suggestion
6. Define two [`http:Client`](https://ballerina.io/learn/by-example/#http-client) objects to send requests to the backend services and one [`email:SmtpClient`](https://ballerina.io/learn/by-example/#email-client) client to send emails.
```

3. Generate record types corresponding to the payloads from the hospital and payment backend services by providing samples of the expected JSON payloads.

    The payload from the hospital backend service will be a JSON object similar to the following.

    ```json
    {
        "appointmentNumber": 1,
        "doctor": {
            "name": "thomas collins",
            "hospital": "grand oak community hospital",
            "category": "surgery",
            "availability": "9.00 a.m - 11.00 a.m",
            "fee": 7000.0
        },
        "patient": {
            "name": "John Doe",
            "dob": "1940-03-19",
            "ssn": "234-23-525",
            "address": "California",
            "phone": "8770586755",
            "email": "johndoe@gmail.com"
        },
        "hospital": "grand oak community hospital",
        "confirmed": false,
        "appointmentDate": "2023-10-02"
    }
    ```

    ![Define records](/learn/images/integration-tutorials/sending-emails-from-a-service/define_records.gif)

    The generated records will be as follows.

    ```ballerina

>     type Patient record {|

Reviewer 2: Have to be open?

Author: Yes and might have retake the gifs

Author: Yes and might have to retake some of the gifs

    $ bal new sending-emails-from-a-service

>     ```

Reviewer 2: L65

> Call the payment backend service to make the payment and retrieve the reservation response, which will have a payload similar to that shown below. If the payment is successful, send an email to the user with the appointment details.

Shall we say

> Call the payment backend service to make the payment to confirm the appointment. If the payment is successful, the response payload will be similar to that shown below. 

Then after the JSON sample, add

> 4. If the payment is successful, send an email to the user with the appointment details.

        The generated service will be as follows.

        ```ballerina
        service /healthcare on new http:Listener(8290) {
            resource function post categories/[string category]/reserve(ReservationRequest payload) 
                    returns http:Created|http:NotFound|http:InternalServerError {
                
            }
        }
        ```

5. Define [configurable variables](https://ballerina.io/learn/by-example/#configurability) for the URLs of the backend services. Also, define configurable variables for the host, username, and password of the SMTP client.

    ![Define a configurable variable](/learn/images/integration-tutorials/sending-emails-from-a-service/define_a_configurable_variable.gif)

    The generated code will be as follows.

    ```ballerina
    configurable string hospitalServicesBackend = "http://localhost:9090";
    configurable string paymentBackend = "http://localhost:9090/healthcare/payments";
    configurable string host = "smtp.gmail.com";
    configurable string username = ?;
    configurable string password = ?;
    ```

>     > **Note:** Enable two factor authentication on your Google account, generate an app password, and use the app password in place of your email password. The app password can be generated with [this link](https://myaccount.google.com/apppasswords?pli=1&rapt=AEjHL4Mf5XDD4rE79YJpP5E2NoNwhvXMET_TWyBcQRn-HMzt0PI8BmptpMGRiBVIamW-0ECgVZtXxMRA19bL4Wfnq_hmjBEMqA).

Reviewer: `This link` link must be `https://myaccount.google.com/apppasswords`

Author: Fixed with https://github.com/ballerina-platform/ballerina-dev-website/pull/8236

>     > **Note:** Enable two factor authentication on your Google account, generate an app password, and use the app password in place of your email password. The app password can be generated with [this link](https://myaccount.google.com/apppasswords?pli=1&rapt=AEjHL4Mf5XDD4rE79YJpP5E2NoNwhvXMET_TWyBcQRn-HMzt0PI8BmptpMGRiBVIamW-0ECgVZtXxMRA19bL4Wfnq_hmjBEMqA).

Reviewer: Using `Config.toml` part is not explained. IMO, the user can't find a way to use them in the tutorial.

    The generated code will be as follows.

    ```ballerina
    configurable string hospitalServicesBackend = "http://localhost:9090";
    configurable string paymentBackend = "http://localhost:9090/healthcare/payments";
    configurable string host = "smtp.gmail.com";
    configurable string username = ?;
    configurable string password = ?;
    ```

    > **Note:** Enable two factor authentication on your Google account, generate an app password, and use the app password in place of your email password. The app password can be generated with [this link](https://myaccount.google.com/apppasswords?pli=1&rapt=AEjHL4Mf5XDD4rE79YJpP5E2NoNwhvXMET_TWyBcQRn-HMzt0PI8BmptpMGRiBVIamW-0ECgVZtXxMRA19bL4Wfnq_hmjBEMqA).

6. Define two [`http:Client`](https://ballerina.io/learn/by-example/#http-client) objects to send requests to the backend services and one [`email:SmtpClient`](https://ballerina.io/learn/by-example/#email-client) client to send emails.
   
   ![Define the clients](/learn/images/integration-tutorials/sending-emails-from-a-service/define_clients.gif)

   The generated code will be as follows.

   ```ballerina

>    final http:Client hospitalServicesEP = check new(hospitalServicesBackend);

Reviewer 2: Aren't these added with a space between new and ( ?

    |};
    ```

    > **Note:**
    > While it is possible to work with the JSON payload directly, using record types offers several advantages including enhanced type safety, data validation, and better tooling experience (e.g., completion).

> 4. Define the [HTTP service (REST API)](https://ballerina.io/learn/by-example/#rest-service) that has the resource that accepts user requests, retrieves relevant details from the backend service, and responds to the request.

Reviewer 2: Doesn't mention sending emails?

    - Open the [Ballerina HTTP API Designer](https://wso2.com/ballerina/vscode/docs/design-the-services/http-api-designer) in VS Code.

    - Use `/healthcare` as the service path (or the context) for the service attached to the listener that is listening on port `8290`.

>         ![Define the service](/learn/images/integration-tutorials/sending-emails-from-a-service/define_a_service.gif)

Reviewer 2: This gif still seems to have old code?

Reviewer 2: Same for the subsequent ones.

                
            }
        }
        ```

5. Define [configurable variables](https://ballerina.io/learn/by-example/#configurability) for the URLs of the backend services. Also, define configurable variables for the host, username, and password of the SMTP client.

    ![Define a configurable variable](/learn/images/integration-tutorials/sending-emails-from-a-service/define_a_configurable_variable.gif)

    The generated code will be as follows.

    ```ballerina
    configurable string hospitalServicesBackend = "http://localhost:9090";
    configurable string paymentBackend = "http://localhost:9090/healthcare/payments";
    configurable string host = "smtp.gmail.com";
    configurable string username = ?;
    configurable string password = ?;
    ```

    > **Note:** Enable two factor authentication on your Google account, generate an app password, and use the app password in place of your email password. The app password can be generated with [this link](https://myaccount.google.com/apppasswords?pli=1&rapt=AEjHL4Mf5XDD4rE79YJpP5E2NoNwhvXMET_TWyBcQRn-HMzt0PI8BmptpMGRiBVIamW-0ECgVZtXxMRA19bL4Wfnq_hmjBEMqA).

> 6. Define two [`http:Client`](https://ballerina.io/learn/by-example/#http-client) objects to send requests to the backend services and one [`email:SmtpClient`](https://ballerina.io/learn/by-example/#email-client) client to send emails.

Reviewer 2: ```suggestion
6. Define two [`http:Client`](https://ballerina.io/learn/by-example/#http-client) objects to send requests to the backend services and one [`email:SmtpClient`](https://ballerina.io/learn/by-example/#email-client) object to send emails.
```

            log:printDebug("Email sent successfully",
                            name = patient.name,
                            appointmentNumber = appointmentNumber);
            return <http:Created>{};
        }
    }

> 

Reviewer 2: L293 onward, we decided to remove repeating the code snippets when explaining them.

Author: If we remove the code snippets, wouldn't it be harder to understand the descriptions?

Reviewer 2: Since we describe in the order of the flow, should be OK without IMO.

   - Use the `is` check to decide the flow based on the response to the client call. If the request failed, return an `http:NotFound` response. Else, if the payload could not be bound to `Appointment` as expected or if there were any other failures, respond with an `http:InternalServerError` response.

   - If the appointment reservation was successful, we can make the payment by making a `POST` request to the payment service. The payload includes details extracted out from the original request (for `card_number`) and the appointment reservation response (for `appointmentNumber`, `doctor`, `patient`, `fee`, and `confirmed`).

   - If the payment was successful, the next and final step is to send an email to the user with the appointment details. We send the email specifying the user's email address, the subject, and the email body.

   - If the email is sent successfully, the response will be an `http:Created` response. If the email sending process resulted in an error, an `http:InternalServerError` response will be returned.

You have successfully developed the required service.

> #### Complete source

Reviewer 2: Doesn't compile.

>             log:printDebug("Email sent successfully",

Reviewer 2: Remove log statements except error logs.

   - The first backend call is a `POST` request to the hospital service to reserve the appointment. The `hospital_id` and `category` values are used as path parameters.

>    - Use the `is` check to decide the flow based on the response to the client call. If the request failed, return an `http:NotFound` response. Else, if the payload could not be bound to `Appointment` as expected or if there were any other failures, respond with an `http:InternalServerError` response.

Reviewer 2: Shall we elaborate here to say

```suggestion
   - Use the `is` check to decide the flow based on the response to the client call. If the request failed with a `4xx` status code, return an `http:NotFound` response. Else, if the payload could not be bound to `Appointment` as expected or if there were any other failures, respond with an `http:InternalServerError` response.
```

    decimal discounted;
    string paymentID;
    string status;
};

> type PatientWithCardNo record {

Reviewer: @MaryamZi So we are not going to have closed records in the complete source code as well right?

Reviewer 2: No. What I meant is that when we say this is the generated code, it has to be the generated code. We can then close it and include a note and say we can make these closed records and give the reason.

    string status;
|};

configurable string hospitalServicesBackend = "http://localhost:9090";
configurable string paymentBackend = "http://localhost:9090/healthcare/payments";
configurable string host = "smtp.gmail.com";
configurable string username = ?;
configurable string password = ?;

final http:Client hospitalServicesEP = check new(hospitalServicesBackend);
final http:Client paymentEP = check new(paymentBackend);
final email:SmtpClient smtpClient = check new(host, username, password);

function initializeHttpClient(string url) returns http:Client|error => new (url);

> function initializeEmailClient() returns email:SmtpClient|error => new (host, username, password);

Reviewer 2: These should be removed.

service /healthcare on new http:Listener(8290) {

    resource function post categories/[string category]/reserve(ReservationRequest payload)
            returns http:Created|http:NotFound|http:InternalServerError {

        ReservationRequest {

>             patient,

Reviewer 2: Need to remove destructuring.

### Step 3: Build and run the service

> ![Run the service](/learn/images/integration-tutorials/sending-emails-from-a-service/run_the_service.gif)

Reviewer 2: At least for others let's follow /Users/username/integration-tutorials/tutorial-name as the root.

Reviewer: +1

> **Note:**
> Alternatively, you can run this service by navigating to the project root and using the `bal run` command.
>
> ```
> sending-a-message-to-a-service$ bal run
> Compiling source
>         integration_tutorials/sending_a_message_to_a_service:0.1.0
>
> Running executable
> ```

### Step 4: Try out the use case

Let's test the use case by sending a request to the service.

#### Start the backend service

Download the JAR file for the [backend service](https://github.com/ballerina-guides/integration-tutorials/blob/main/backends/hospital-service/hospitalservice.jar) and execute the following command to start the service.

```

> bal run hospitalservice.jar

Reviewer 2: ```suggestion
$ bal run hospitalservice.jar
```

Reviewer: Check other bash scripts too

function initializeEmailClient() returns email:SmtpClient|error => new (host, username, password);

> service /healthcare on new http:Listener(port) {

Reviewer: Not related to this line. 

```
function initializeHttpClient(string url) returns http:Client|error => new (url);
function initializeEmailClient() returns email:SmtpClient|error => new (host, username, password);
```

We have decided to remove the function approach to initialize the clients (hospitalServiceBackend, paymentBackend).

    resource function post categories/[string category]/reserve(ReservationRequest payload)
            returns http:Created|http:NotFound|http:InternalServerError {

        ReservationRequest {

>             patient,

Reviewer 2: Why do we need to change the source here?

Reviewer: I think this is due to the bug right? This causes errors in "diagram view" and "try-out".

Author: Yes this change was added to fix the error in `diagram view` and `try-out`

### Step 3: Build and run the service

![Run the service](/learn/images/integration-tutorials/sending-emails-from-a-service/run_the_service.gif)

> **Note:**
> Alternatively, you can run this service by navigating to the project root and using the `bal run` command.
>
> ```
> sending-a-message-to-a-service$ bal run
> Compiling source
>         integration_tutorials/sending_a_message_to_a_service:0.1.0
>
> Running executable
> ```

### Step 4: Try out the use case

Let's test the use case by sending a request to the service.

#### Start the backend service

Download the JAR file for the [backend service](https://github.com/ballerina-guides/integration-tutorials/blob/main/backends/hospital-service/hospitalservice.jar) and execute the following command to start the service.

```
$ bal run hospitalservice.jar
```

#### Send a request

> Use the [Try it](https://wso2.com/ballerina/vscode/docs/try-the-services/try-http-services/) feature to send a request to the service. Specify the port used in the service and use `surgery` as the path parameter and send a request with a JSON payload in the following form.

Reviewer 2: ```suggestion
Use the [Try it](https://wso2.com/ballerina/vscode/docs/try-the-services/try-http-services/) feature to send a request to the service. Specify the port used in the service and use `surgery` as the path parameter. Use the following as the request payload.
```

>     ```json

Reviewer 2: Doesn't render properly.

### Step 3: Build and run the service

![Run the service](/learn/images/integration-tutorials/sending-emails-from-a-service/run_the_service.gif)

> **Note:**
> Alternatively, you can run this service by navigating to the project root and using the `bal run` command.
>
> ```
> sending-a-message-to-a-service$ bal run
> Compiling source
>         integration_tutorials/sending_a_message_to_a_service:0.1.0
>
> Running executable
> ```

### Step 4: Try out the use case

Let's test the use case by sending a request to the service.

#### Start the backend service

Download the JAR file for the [backend service](https://github.com/ballerina-guides/integration-tutorials/blob/main/backends/hospital-service/hospitalservice.jar) and execute the following command to start the service.

```
$ bal run hospitalservice.jar
```

#### Send a request

Use the [Try it](https://wso2.com/ballerina/vscode/docs/try-the-services/try-http-services/) feature to send a request to the service. Specify the port used in the service and use `surgery` as the path parameter.

> 

Reviewer 2: Since this is a POST request, we have to include the sample request payload here.

### Step 3: Build and run the service

![Run the service](/learn/images/integration-tutorials/sending-emails-from-a-service/run_the_service.gif)

> **Note:**
> Alternatively, you can run this service by navigating to the project root and using the `bal run` command.
>
> ```
> sending-a-message-to-a-service$ bal run
> Compiling source

> >         integration_tutorials/sending_a_message_to_a_service:0.1.0

Reviewer 2: Wrong package name.
